syntax = "proto2";

// Protocol Version 7

package org.oscim.database.oscimap4;

message Data {
	message Element {
		
		// number of geometry 'indices'
		optional uint32 num_indices     = 1 [default = 1];
		
		// number of 'tags'
		optional uint32 num_tags        = 2 [default = 1];

		// elevation per coordinate 
		// (pixel relative to ground meters)
		// optional bool has_elevation     = 3 [default = false];

		optional uint64 id				= 4;

		optional uint64 reserved5       = 5;

		optional uint32 reserved6       = 6;

		optional sint32 reserved7       = 7;

		optional bool reserved8         = 8;

		optional string reserved9       = 9;

		// reference to tile.tags
		repeated uint32 tags            = 11 [packed = true];
		
		// A list of number of coordinates for each geometry.
		// - polygons are separated by one '0' index
		// - for single points this can be omitted.
		// e.g 2,2 for two lines with two points each, or
		// 4,3,0,4,3 for two polygons with four points in
		// the outer ring and 3 points in the inner.
		
		repeated uint32 indices         = 12 [packed = true];
		
		// single delta encoded coordinate x,y pairs scaled
		// to a tile size of 4096
		// note: geometries start at x,y = tile size / 2
		
		repeated sint32 coordinates     = 13 [packed = true];
		
		//---------------- optional items ---------------
		// osm layer [-5 .. 5] -> [0 .. 10]
		optional uint32 layer           = 21 [default = 5];

		// area in map units
		optional uint64 area            = 23;

		// label position
		repeated sint32 label			= 31 [packed = true];
		
		// amenity kind
		optional uint32 kind            = 32;
		
		// amenity type
		// used when element is not only amenity but also
		// something else, e.g. building
		optional uint32 type            = 22;

		// elevation in meters
		optional sint32 elevation       = 33;
		
		// depth, precision 1/100m
		optional sint32 depth           = 24;

		// building height, precision 1/100m
		optional sint32 height          = 34;

		// building height, precision 1/100m
		optional sint32 min_height      = 35;

		// building walls color
		optional uint32 building_color  = 36;

		// building roof color
		optional uint32 roof_color      = 37;

		// building roof height, precision 1/100m
		optional sint32 roof_height     = 39;

		// building roof shape
		optional sint32 roof_shape      = 40;

		// building roof direction, degrees * 10
		optional sint32 roof_direction  = 41;

		// building roof orientation, along - false, across - true
		optional bool roof_across       = 42 [default = false];

		// building addr:housenumber
		optional string housenumber     = 38;
	}
	
	required uint32 version   = 1;
	
	// tile creation time
	optional uint64 timestamp = 2;

	// tile is completely water (not used yet)
	optional bool water       = 3;
	
	// number of 'tags'
	required uint32 num_tags  = 11;
	optional uint32 num_keys  = 12 [default = 0];
	optional uint32 num_vals  = 13 [default = 0];

	// strings referenced by tags
	repeated string keys      = 14;
	// separate common attributes from label to
	// allow
	repeated string values    = 15;
	
	// (key[0xfffffffc] | type[0x03]), value pairs
	// key: uint32       -> reference to key-strings
	// type 0: attribute -> uint32 reference to value-strings
	// type 1: string    -> uint32 reference to label-strings
	// type 2: sint32
	// type 3: float
	// value: uint32 interpreted according to 'type'
		
	repeated uint32 tags      = 16 [packed = true];
	
	
	// linestrings
	repeated Element lines    = 21;

	// polygons (MUST be implicitly closed) 
	repeated Element polygons = 22;

	// points
	repeated Element points   = 23;
}
